### PR TITLE
[fix](logstash plugin) fix group_commit and sprintf with missing value

### PR DESCRIPTION
**Pull request by joker-star-l**
fix: dev lib version
fix: default map to empty string
fix: do not retry if label duplication

### CODE REVIEW

**Comment by joker-star-l** [comment]
run buildall

File: extension/logstash/lib/logstash/outputs/doris.rb [hunk 1]
```diff
@@ -90,17 +90,17 @@ def print_plugin_info()
    end
 
    def register
-      @http_query = "/api/#{db}/#{table}/_stream_load"
+      @http_query = "/api/#{@db}/#{@table}/_stream_load"
```
Comment (xiaokang) [inline extension/logstash/lib/logstash/outputs/doris.rb:93]: Is '@' necessary?

File: extension/logstash/lib/logstash/outputs/doris.rb [hunk 2]
```diff
@@ -90,17 +90,17 @@ def print_plugin_info()
...
-      parse_http_hosts(http_hosts,
+      parse_http_hosts(@http_hosts,
       ShortNameResolver.new(ttl: @host_resolve_ttl_sec, logger: @logger))
 
       @request_headers = make_request_headers
       @logger.info("request headers: ", @request_headers)
 
       @group_commit = false
-      if http_headers.has_key?("group_commit") && http_headers["group_commit"] != "off_mode"
+      if @request_headers.has_key?("group_commit") && @request_headers["group_commit"] != "off_mode"
```
Comment (xiaokang) [inline extension/logstash/lib/logstash/outputs/doris.rb:103]: Is it a mistake to use http_headers?

File: extension/logstash/lib/logstash/outputs/doris.rb [hunk 3]
```diff
@@ -211,7 +211,15 @@ def send_events(events)
          rescue => e
             @logger.warn("doris stream load response: #{response} is not a valid JSON")
          end
-         if response_json["Status"] == "Success"
+
+         status = response_json["Status"]
+
+         if status == 'Label Already Exists'
+           @logger.warn("Label already exists: #{response_json['Label']}")
+           break
```
Comment (xiaokang) [inline extension/logstash/lib/logstash/outputs/doris.rb:219]: What's the effect of break;

File: extension/logstash/lib/logstash/outputs/doris.rb [hunk 4]
```diff
@@ -252,13 +260,13 @@ def make_request(documents, http_headers, hosts, query, host = "")
 
       response = ""
       begin
-         response = RestClient.put(url, documents, http_headers) { |response, request, result|
-                case response.code
+         response = RestClient.put(url, documents, http_headers) { |res, request, result|
```
Comment (xiaokang) [inline extension/logstash/lib/logstash/outputs/doris.rb:263]: Why change response to res?

File: extension/logstash/logstash-output-doris.gemspec [hunk 5]
```diff
@@ -41,7 +41,7 @@ Gem::Specification.new do |s|
   s.add_runtime_dependency 'mini_cache', ">= 1.0.0", "< 2.0.0"
   s.add_runtime_dependency "rest-client", '~> 2.1'
 
-  s.add_development_dependency 'logstash-devutils', '~> 2.0', '>= 2.0.3'
-  s.add_development_dependency 'sinatra', '~> 2.0', '>= 2.0.8.1'
-  s.add_development_dependency 'webrick', '~> 1.6'
+  s.add_development_dependency 'logstash-devutils', '~> 1.3'
```
Comment (xiaokang) [inline extension/logstash/logstash-output-doris.gemspec:44]: Why change version?

File: extension/logstash/lib/logstash/util/formater.rb [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+require 'java'
+
+class Formater
+
+  # copied form https://www.rubydoc.info/gems/logstash-event/LogStash/Event#sprintf-instance_method
+  # modified by doris
```
Comment (xiaokang) [inline extension/logstash/lib/logstash/util/formater.rb:23]: explain the modified content and reason.

File: extension/logstash/lib/logstash/outputs/doris.rb (hunk 3 above)
Comment (xiaokang) [reply extension/logstash/lib/logstash/outputs/doris.rb:219]: It's more safe to change the lable and retry.

File: extension/logstash/lib/logstash/outputs/doris.rb [hunk 7]
```diff
@@ -77,9 +77,9 @@ class LogStash::Outputs::Doris < LogStash::Outputs::Base
 
 
    def print_plugin_info()
-      @@plugins = Gem::Specification.find_all{|spec| spec.name =~ /logstash-output-doris/ }
-      @plugin_name = @@plugins[0].name
-      @plugin_version = @@plugins[0].version
+      @plugins = Gem::Specification.find_all{|spec| spec.name =~ /logstash-output-doris/ }
```
Comment (xiaokang) [inline extension/logstash/lib/logstash/outputs/doris.rb:80]: refer to http output plugin

File: extension/logstash/lib/logstash/outputs/doris.rb (hunk 1 above)
Comment (joker-star-l) [reply extension/logstash/lib/logstash/outputs/doris.rb:93]: It's not necessary, but it's usually added in logstash plugin.

File: extension/logstash/lib/logstash/outputs/doris.rb (hunk 2 above)
Comment (joker-star-l) [reply extension/logstash/lib/logstash/outputs/doris.rb:103]: Yes.

File: extension/logstash/lib/logstash/outputs/doris.rb (hunk 4 above)
Comment (joker-star-l) [reply extension/logstash/lib/logstash/outputs/doris.rb:263]: `response` was defined earlier.

File: extension/logstash/logstash-output-doris.gemspec (hunk 5 above)
Comment (joker-star-l) [reply extension/logstash/logstash-output-doris.gemspec:44]: This is the development environment version of the library and does not affect the runtime environment. The previous version had dependency conflicts.

File: extension/logstash/lib/logstash/outputs/doris.rb (hunk 7 above)
Comment (joker-star-l) [reply extension/logstash/lib/logstash/outputs/doris.rb:80]: The http plugin has no code for this.

**Comment by joker-star-l** [comment]
run buildall

**Review by xiaokang — APPROVED** [review]
LGTM

**Review by qidaye — APPROVED** [review]
LGTM

**Review by echo-hhj — APPROVED** [review]
LGTM

**Pull request merged**